src/NServiceBus.Core/Routing/DistributionContext.cs

```diff
@@ -0,0 +1,49 @@
+namespace NServiceBus.Routing
+{
+    using System.Collections.Generic;
+    using Extensibility;
+    using Pipeline;
+
+    /// <summary>
+    /// The distribution context for the DistributionStrategy implementers.
+    /// </summary>
+    public class DistributionContext
+    {
+        /// <summary>
+        /// Creates a new distribution context.
+        /// </summary>
+        public DistributionContext(string[] receiverAddresses, OutgoingLogicalMessage message, string messageId, Dictionary<string, string> headers, ContextBag context)
+        {
+            ReceiverAddresses = receiverAddresses;
+            Message = message;
+            MessageId = messageId;
+            Headers = headers;
+            Context = context;
+        }
+
+        /// <summary>
+        /// The receiver addresses that can be taken into account for distribution.
+        /// </summary>
+        public string[] ReceiverAddresses { get; }
+
+        /// <summary>
+        /// The id of the outgoing message.
+        /// </summary>
+        public string MessageId { get; private set; }
+
+        /// <summary>
+        /// The headers of the outgoing message.
+        /// </summary>
+        public Dictionary<string, string> Headers { get; }
```

Comment:
but allowing the headers to be mutable will also locking us to certain phases. Not sure is an issue tbh but allocations in case you really access headers isn't my main concern. Maybe I'm overthinking it, no strong opinion. Shame that readonly collections are such a PITA.